app/src/main/java/nodomain/freeyourgadget/gadgetbridge/service/devices/miscale2/MiScale2DeviceSupport.java

```diff
@@ -0,0 +1,290 @@
+import nodomain.freeyourgadget.gadgetbridge.service.btle.profiles.IntentListener;
+import nodomain.freeyourgadget.gadgetbridge.service.btle.profiles.deviceinfo.DeviceInfoProfile;
+import nodomain.freeyourgadget.gadgetbridge.util.GB;
+
+public class MiScale2DeviceSupport extends AbstractBTLEDeviceSupport {
+
+    private static final Logger LOG = LoggerFactory.getLogger(MiScale2DeviceSupport.class);
+    private static final String UNIT_KG = "kg";
+    private static final String UNIT_LBS = "lb";
+    private static final String UNIT_JIN = "jīn";
+    private final DeviceInfoProfile<MiScale2DeviceSupport> deviceInfoProfile;
+    private final GBDeviceEventVersionInfo versionCmd = new GBDeviceEventVersionInfo();
+    private final IntentListener mListener = new IntentListener() {
+        @Override
+        public void notify(Intent intent) {
+            String s = intent.getAction();
+            if (s.equals(DeviceInfoProfile.ACTION_DEVICE_INFO)) {
+                handleDeviceInfo((nodomain.freeyourgadget.gadgetbridge.service.btle.profiles.deviceinfo.DeviceInfo) intent.getParcelableExtra(DeviceInfoProfile.EXTRA_DEVICE_INFO));
+            }
+        }
+    };
+
+    public MiScale2DeviceSupport() {
+        super(LOG);
+        addSupportedService(GattService.UUID_SERVICE_GENERIC_ACCESS);
+        addSupportedService(GattService.UUID_SERVICE_GENERIC_ATTRIBUTE);
+        addSupportedService(GattService.UUID_SERVICE_DEVICE_INFORMATION);
+        addSupportedService(GattService.UUID_SERVICE_BODY_COMPOSITION);
+        addSupportedService(UUID.fromString("00001530-0000-3512-2118-0009af100700"));
+
+        deviceInfoProfile = new DeviceInfoProfile<>(this);
+        deviceInfoProfile.addListener(mListener);
+        addSupportedProfile(deviceInfoProfile);
+    }
+
+    @Override
+    protected TransactionBuilder initializeDevice(TransactionBuilder builder) {
+        builder.add(new SetDeviceStateAction(getDevice(), GBDevice.State.INITIALIZING, getContext()));
+
+        LOG.debug("Requesting Device Info!");
+        deviceInfoProfile.requestDeviceInfo(builder);
+        builder.add(new SetDeviceStateAction(getDevice(), GBDevice.State.INITIALIZED, getContext()));
+
+        // Weight and body composition
+        builder.setGattCallback(this);
+        builder.notify(getCharacteristic(GattCharacteristic.UUID_CHARACTERISTIC_BODY_COMPOSITION_MEASUREMENT), true);
+
+        return builder;
+    }
+
+    @Override
+    public boolean onCharacteristicChanged(BluetoothGatt gatt, BluetoothGattCharacteristic characteristic) {
+        super.onCharacteristicChanged(gatt, characteristic);
+
+        UUID characteristicUUID = characteristic.getUuid();
+        if (characteristicUUID.equals(GattCharacteristic.UUID_CHARACTERISTIC_BODY_COMPOSITION_MEASUREMENT)) {
+            final byte[] data = characteristic.getValue();
+
+            boolean stabilized = testBit(data[1], 5) && !testBit(data[1], 7);
```

Comment:
Does this implementation conform to the standard BLE profile/service for this? If so, we could even implement it generically like the DeviceInfoProfile, so that it can be reused for other devices.

Reply:
Unfortunately, no, it is not standard.
Here is the spec:
https://www.bluetooth.com/specifications/gatt/viewer?attributeXmlFile=org.bluetooth.characteristic.body_composition_measurement.xml

The order of the data is quite different